### DISCUSSION TITLE
EF Core Migrations: Design-time

### DISCUSSION THREAD

**Discussion by bricelam** [bricelam/bricelam.github.io#89] [General]
# EF Core Migrations: Design-time

In previous posts, I’ve talked the new .NET Core CLI commands and the enhanced NuGet commands in Entity Framework Core. This post dives into some of the changes we’ve made to enhance the design-time experience of Migrations.

https://www.bricelam.net/2014/12/16/ef7-migrations-designtime.html

**Comment by bricelam** [comment]
_From **[Igor Oleinikov](https://disqus.com/by/igoroleinikov)** on December 18, 2014 17:12_

<p>Exciting,,,<br>How the migrations look like? Is it still just a code? To resolve conflicts is it required to apply other changes first?<br>Thanks Brice!</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on December 18, 2014 17:17_

<p>They don't look too different. I want to cover the actual API changes in one last blog post on the subject. We're still working out the format of the model snapshot, our current thinking is that it'll end up being JSON-based (right now, it uses a not-so-strong-typed version of the Fluent API).</p><p>No, in most cases you shouldn't need to apply (or unapply) migrations in order to resolve the conflicts.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Igor Oleinikov](https://disqus.com/by/igoroleinikov)** on December 18, 2014 17:22_

<p>Thanks, it's really awesome. Looking forward to try it.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Julie Lerman](https://disqus.com/by/julielerman)** on January 21, 2015 16:45_

<p>Hey Brice, based on this, I would expect add-migration to not need the connection string but I am still being forced to supply it  Is that correct or have I missed something? Thanks</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on January 25, 2015 5:50_

<p>It may be a bug. I suspect we're eagerly getting the list of applied migrations even though we don't use it. I'm doing a big cleanup of the migrations runtime at the moment. I'll be sure to look into this issue too.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Julie Lerman](https://disqus.com/by/julielerman)** on January 25, 2015 14:12_

<p>Do you want me to file an issue?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on January 25, 2015 17:46_

<p>Yeah, that's a good way to make sure I don't forget. ;)</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Julie Lerman](https://disqus.com/by/julielerman)** on January 26, 2015 15:24_

<p>done</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[David](https://disqus.com/by/damccull)** on January 26, 2015 0:29_

<p>Do these commands already exist in the CTP5 of visual studio if we start a new MVC6 project? If not, are they able to be added somehow?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on January 26, 2015 0:36_

<p>Yes, I beliv they are included in the starter template.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[David](https://disqus.com/by/damccull)** on January 26, 2015 3:06_

<p>Hmm. I see the dependency there. I am unable to use these commands, however. I have installed KVM according to the instructions at <a href="http://www.asp.net/vnext/overview/aspnet-vnext/vc#efm" rel="nofollow noopener" title="http://www.asp.net/vnext/overview/aspnet-vnext/vc#efm">http://www.asp.net/vnext/ov...</a> and I can use 'k ef' and its subcommands, but not 'add-migration' or similar. Also, the way I'm doing it doesn't seem to support removing a migration...suggestions?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on January 26, 2015 5:34_

<p>PowerShell is not available in ASP.NET 5 projects. (See <a href="https://github.com/aspnet/XRE/issues/952" rel="nofollow noopener" title="https://github.com/aspnet/XRE/issues/952">aspnet/XRE#952</a>.) There are equivalent 'k ef' commands you can use instead.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[David](https://disqus.com/by/damccull)** on January 26, 2015 6:19_

<p>Thanks for pointing this out. I hope that we'll be able to use the powershell commands in the future. Keep up the good work. We all appreciate it.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[togakangaroo](https://disqus.com/by/togakangaroo)** on March 4, 2015 3:48_

<p>So this is all good.</p><p>I wonder though if you guys are putting any effort toward multiple databases and especially polyglot persistence scenarios.</p><p>Frequently, the reason someone is removing a table from Sql Server is because they're shifting that data into Postgres, CouchDb, or Redis. I realize that you guys are *just* Entity Framework, but are you baking in the hooks necessary to wrap EF migrations in a more general infrastructure migrations framework?</p><p>Last time I took a look (Admittedly, EF 4) this was impossible.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on March 4, 2015 17:14_

<p>EF7 is all about <a href="http://blogs.msdn.com/b/adonet/archive/2014/05/19/ef7-new-platforms-new-data-stores.aspx" rel="nofollow noopener" title="http://blogs.msdn.com/b/adonet/archive/2014/05/19/ef7-new-platforms-new-data-stores.aspx">targeting new data stores</a> including non-relational databases. However, Migrations is still primarily about relational database schema migrations. It's conceivable, however, that you could use EF7 to do cross-store data migrations by pumping data from one DbContext into another. We do want to look at enabling relational data migrations in the future, but we haven't thought much about cross-store scenarios.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Kasey Krehbiel](https://disqus.com/by/kaseykrehbiel)** on May 12, 2015 15:20_

<p>What kind of work would have to be done to migrate from EF 6 Migrations to EF 7 Migrations?</p><p>Also, what sort of work would it take to generate EF 7 Migrations from an existing solution?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on May 12, 2015 15:43_

<p>The APIs are pretty different; I don't think we'll support converting existing migrations. Instead, you'll probably "start fresh" after porting you application to EF7.</p><p>We've only briefly thought about it the scenario. Follow issue <a href="https://github.com/aspnet/EntityFramework/issues/2167" rel="nofollow noopener" title="https://github.com/aspnet/EntityFramework/issues/2167">#2167</a> to watch it evolve.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Rainzo](https://disqus.com/by/Rainzo)** on June 1, 2015 20:53_

<p>The add-migration gives the error that enable-migrations gave...has there been another change?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Sergiy Korzh](https://disqus.com/by/sergiykorzh)** on August 22, 2015 6:46_

<p>Migration are good to move between two versions of DB when your project is in production already.<br>But for initial stage when DB is changed literally each day it's better to drop the whole DB at once and fill it with some testing data on start. Something like DropCreateDatabaseIfModelChanges option in previous versions of EF.<br>Is it possible to implement such scenario in EF7?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on August 22, 2015 18:29_

<p>Yes! It is a perfectly valid workflow for rapid prototyping and testing. We have <code>DbContext.Database.EnsureCreated()</code> and <code>EnsureDeleted()</code> to help in these scenarios.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Sergiy Korzh](https://disqus.com/by/sergiykorzh)** on August 23, 2015 7:46_

<p>Great. Thank you.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Shimmy](https://disqus.com/by/weitzhandler)** on September 9, 2015 10:54_

<p>How about database initialization? Is there an out-the-box way or conventional way to achieve this with EF7?</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on September 9, 2015 15:50_

<p>Most of the building blocks are there to do it yourself. The big piece missing is detecting "on model changed", but you might be able to do it using reverse engineer and the model differ, or by manually maintaining a hash (or snapshot) of the model in the database.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Shimmy](https://disqus.com/by/weitzhandler)** on September 18, 2015 3:20_

<p>Thanks for your reply, but I feel sorry DB initialization is gone / doesn't have a standardized way. DB seeding was very important to me.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **Stefan Babos** on March 10, 2016 7:49_

<p>I have the same problem. I think one can use SQL if override Migration.Up migrationBuilder.Sql(...); But on the other hand I think the sql commands are not primary used in EF. I'd rather use Add and SaveChanges - but I have not found this. Indeed, part of the data migration to the new version of data often will need to edit the data and not just add or remove the column in the table.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Perumal Samy](https://disqus.com/by/disqus_QCsAEbmKEV)** on May 17, 2017 8:18_

<p>Hi Brice, <br>I could not find how to check whether new migration needed or not using c# when I go for deployment.<br>Can you suggest me how to implement my following requirement?</p><p>1. I always want to update my database when I deploy new version of my <a href="http://asp.net" rel="nofollow noopener" title="asp.net">asp.net</a> core web application<br>2. I made it by calling context.Database.Migrate() method in startup class.<br>3. All are working fine.<br>4. I want to generate migrations automatcially via genkins<br>5. Genkins will fire a command or call an exe<br>6. Through command or c# exe, I should be able to verify whether there is a need to generate migration file or not.<br>7. if needed, I will generate the migration file and check in it using genkins<br>8. otherwise, I will just use the code to generate MSI as it is</p><p>Thanks In advance,</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Reply by bricelam** [reply]
_From **[Brice Lambson](https://disqus.com/by/bricelam)** on May 17, 2017 15:00_

<p>See my Stack Overflow answer on <a href="http://stackoverflow.com/a/40871947/475031" rel="nofollow noopener" title="http://stackoverflow.com/a/40871947/475031">how to programmatically generate migrations</a>.</p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by bricelam** [comment]
_From **[Jared Barneck](https://disqus.com/by/rhyous)** on October 7, 2019 20:03_

<p>I don't agree with this decision. I shared my reasons here: <a href="https://github.com/aspnet/EntityFrameworkCore/issues/6214#issuecomment-539179309" rel="nofollow noopener" title="https://github.com/aspnet/EntityFrameworkCore/issues/6214#issuecomment-539179309">https://github.com/aspnet/E...</a></p>

_Copied from original comment on [Disqus](https://disqus.com/home/forum/bricelam)_

**Comment by mahop-net** [comment]
Just to let you know: Missing Automated Migrations in EF Core is blocking us to use it. We still use old EF 6 in all our Applications. Especially in bigger Teams it's horribly time consuming to create Migrations on design time...
